fundraiser/src/main.sw

```diff
@@ -0,0 +1,505 @@
+contract;
+
+dep abi;
+dep data_structures;
+dep errors;
+dep events;
+dep utils;
+
+// Identity and result importing via * is a workaround until bug is fixed
+use std::{
+    assert::require,
+    block::height,
+    chain::auth::{AuthError, msg_sender},
+    constants::BASE_ASSET_ID,
+    context::{call_frames::msg_asset_id, msg_amount, this_balance},
+    contract_id::ContractId,
+    identity::*,
+    logging::log,
+    result::*,
+    revert::revert,
+    storage::StorageMap,
+    token::transfer,
+};
+
+use abi::Fundraiser;
+use data_structures::{AssetInfo, Campaign, CampaignInfo, Pledge};
+use errors::{CampaignError, CreationError, UserError};
+use events::{CancelledCampaignEvent, ClaimedEvent, CreatedCampaignEvent, PledgedEvent, UnpledgedEvent};
+use utils::{sender_identity, validate_id};
+
+storage {
+    /// Total number of unique assets used across all campaigns
+    asset_count: u64,
+
+    /// Direct look-up for asset data if the user wants to check via a known ID
+    asset_info: StorageMap<ContractId,
+    AssetInfo>, /// O(1) look-up to allow searching via asset_count
+    /// Map(1...asset_count => asset)
+    asset_index: StorageMap<u64,
+    ContractId>, /// The total number of unique campaigns that a user has created
+    /// This should only be incremented
+    /// Cancelling / Claiming should not affect this number
+    campaign_count: StorageMap<Identity,
+    u64>, /// Campaigns that have been created by a user
+
+    /// Map(Identity => Map(1...campaign_count => Campaign)
+    campaign_history: StorageMap<(Identity,
+    u64), Campaign>, /// Data describing the content of a campaign
+    /// Map(Campaign ID => CampaignInfo)
+    campaign_info: StorageMap<u64,
+    CampaignInfo>, /// The total number of unique campaigns that a user has pledged to
+    /// This should only be incremented.
+    /// Unpledging should not affect this number
+    pledge_count: StorageMap<Identity,
+    u64>, /// Record of how much a user has pledged to a specific campaign
+    /// Locked after the deadline
+    /// Map(Identity => Map(1...pledge_count => Pledge))
+    pledge_history: StorageMap<(Identity,
```

Comment:
I don't understand. 
It uses 2 mappings. One is for a look up via a campaign ID and the other is to allow the UI to iterate over the count otherwise the UI needs to constantly have knowledge of the relationship between each campaign and all users